autonomy/deploy/base.py

```diff
@@ -109,7 +108,16 @@ def process_model_args_overrides(self, agent_n: int) -> Dict:
 
     def generate_agents(self) -> List:
         """Generate multiple agent."""
-        return [self.generate_agent(i) for i in range(self.service.number_of_agents)]
+        if self.agent_instances is None:
+            return [
+                self.generate_agent(i) for i in range(self.service.number_of_agents)
+            ]
+
+        agent_override_idx = [
+            (i, self.agent_instances.index(kp["address"]))
+            for i, kp in enumerate(self.keys)
+        ]
```

Comment:
Generate override indexes from `agent_instances` list

Reply:
not very readable and n ^ 2 complexity also. 

https://github.com/valory-xyz/open-autonomy/blob/693915c76e8fe40907a66223687d2f3ebd6228da/autonomy/deploy/base.py#L109-L120

- remove argument `override_idx` from `generate_agent`. It just reduced legibility
- why not sort agent instances as soon as they get set in accordance with the keys? 
- why do we have the keys attribute if this information is already present on the agent_instances?

```
if self.agent_instances: 
    mapping = {kp["address"]: i for i, kp in enumerate(keys)}
    indices = [mapping[x] for x in agent_instances]
else:
    indices = range(len(self.keys))
return list(map(self.generate_agent, indices))
```

Reply:
> why do we have the keys attribute if this information is already present on the agent_instances?

The reason we have both `keys` and `agent_instance` is because the `agent_instances` are the addressed allowed by the service owner to be the part of the deployment and `keys` are defined by the service operator. The `keys` are essentially some funded accounts provided by the service operator to the owner to be a part some service. But they only share the public key or the address of the account, the private keys stays with the service operator, so `agent_instances` are just a list of public keys/account addresses and we need private keys to run the agent instances which will be stored in the `keys.json` file and picked up at the run time. Most of the times the `keys.json` will only hold one account, but we also want to allow a user to run multiple agent instances for a single service so we allow multiple accounts in a single `keys.json` file.

> why not sort agent instances as soon as they get set in accordance with the keys?

This would work if `keys == agent_instances` but it won't be the case unless someone wants to run all n agent instances by themselves.

> remove argument `override_idx` from `generate_agent`. It just reduced legibility

We need that argument because of the situation I explained above, say someone has an addressed which is indexed at position 3 on the `agent_instances` array, In this case we only need to generate 1 agent with the override index being 3. If we go with your logic and just generate `abci3` node instead of `abci0` with overrides being numbered 3 we could face some issues with respective tendermint node.

If the code above seems unnecessarily complicated I can open a ticket to check if your logic affects the tendermint nodes or not if not we can go with your logic.

Reply:
ok just implement the O(n) solution as suggested then and lets move on